Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs

```diff
@@ -0,0 +1,464 @@
+            Coverage.Instrument(s.Tok, "implicit else branch", wr);
+            thenWriter = EmitIf(out guardWriter, false, thenWriter);
+            EmitUnaryExpr(ResolvedUnaryOp.BoolNot, notFalse.E, false, guardWriter, wStmts);
+            TrStmtList(new List<Statement>(), thenWriter);
+          } else {
+            // let's compile the "then" branch
+            wr = EmitIf(out guardWriter, false, wr);
+            EmitExpr(Expression.CreateBoolLiteral(s.Thn.tok, true), false, guardWriter, wStmts);
+            Coverage.Instrument(s.Thn.Tok, "then branch", wr);
+            TrStmtList(s.Thn.Body, wr);
+            Coverage.UnusedInstrumentationPoint(s.Els.Tok, "else branch");
+          }
+        } else {
+          if (s.IsBindingGuard && Options.ForbidNondeterminism) {
+            Error(ErrorId.c_binding_if_forbidden, s.Tok, "binding if statement forbidden by the --enforce-determinism option", wr);
+          }
+
+          var coverageForElse = Coverage.IsRecording && !(s.Els is IfStmt);
+          var thenWriter = EmitIf(out var guardWriter, s.Els != null || coverageForElse, wr);
+          EmitExpr(s.IsBindingGuard ? ((ExistsExpr)s.Guard).AlphaRename("eg_d") : s.Guard, false, guardWriter, wStmts);
+          // We'd like to do "TrStmt(s.Thn, indent)", except we want the scope of any existential variables to come inside the block
+          if (s.IsBindingGuard) {
+            IntroduceAndAssignBoundVars((ExistsExpr)s.Guard, thenWriter);
+          }
+          Coverage.Instrument(s.Thn.Tok, "then branch", thenWriter);
+          TrStmtList(s.Thn.Body, thenWriter);
+
+          if (coverageForElse) {
+            wr = EmitBlock(wr);
+            if (s.Els == null) {
+              Coverage.Instrument(s.Tok, "implicit else branch", wr);
+            } else {
+              Coverage.Instrument(s.Els.Tok, "else branch", wr);
+            }
+          }
+          if (s.Els != null) {
+            TrStmtNonempty(s.Els, wr, wStmts);
+          }
+        }
+
+      } else if (stmt is AlternativeStmt) {
+        var s = (AlternativeStmt)stmt;
+        if (Options.ForbidNondeterminism && 2 <= s.Alternatives.Count) {
+          Error(ErrorId.c_case_based_if_forbidden, s.Tok, "case-based if statement forbidden by the --enforce-determinism option", wr);
+        }
+        foreach (var alternative in s.Alternatives) {
+          var thn = EmitIf(out var guardWriter, true, wr);
+          EmitExpr(alternative.IsBindingGuard ? ((ExistsExpr)alternative.Guard).AlphaRename("eg_d") : alternative.Guard, false, guardWriter, wStmts);
+          if (alternative.IsBindingGuard) {
+            IntroduceAndAssignBoundVars((ExistsExpr)alternative.Guard, thn);
+          }
+          Coverage.Instrument(alternative.Tok, "if-case branch", thn);
+          TrStmtList(alternative.Body, thn);
+        }
+        var wElse = EmitBlock(wr);
+        EmitAbsurd("unreachable alternative", wElse);
+
+      } else if (stmt is WhileStmt) {
```

Comment:
While you're at it and this is a refactoring, could you use your IDE to convert this chain of if-else into a switch statement? It will make debugging much easier.

Reply:
sure. I'll do it in the follow-up, so I don't have to do it twice.